### PR TITLE
Replace gunpoint dataset in tutorials, added OSULeaf dataset

### PR DESCRIPTION
**Pull request by marielledado**
#### Reference Issues/PRs

Issue #294 as part of PyData Amsterdam 2020

#### What does this implement/fix? Explain your changes.
- Replace gunpoint dataset in tutorials 02_classification_univariate, feature_extraction_with_tsfresh, loading_data, rocket, shapelet_transform
- Added OSULeaf dataset to replace gunpoint dataset in shapelet transform tutorial.
- Replaced 01_forecasting tutorial with the same tutorial from PyData Amsterdam code breakfast.
- Removed gunpoint dataset

Notes:
- img folder has also be updated with figures used in the tutorials.
- Added the `random_state` parameter to `train_test_split` for reproducibility.

#### Any other comments?
This is my first PR for sktime and my 3rd time contributing to an open source project! Thanks for having me and for being open to making these changes to the tutorials.

### CODE REVIEW

**Comment by mloning** [comment]
Looks good, thanks for opening the PR! 

Do you want to replace the data set in the other notebooks too as part of this PR? The other notebooks refer to this one for more details on the data set, so I think it makes sense to replace them there as well.

**Comment by marielledado** [comment]
@mloning Yes I'd like to replace the data in the other notebooks as well! I thought about creating a separate branch/PR for each notebook. Happy to put all changes in one PR though, whichever you prefer :)

**Comment by mloning** [comment]
Let's do it all in this PR :)

**Comment by marielledado** [comment]
@mloning so I tried to rename the branch, I guess you can't do that hahahah. I'll continue working on this branch, just ignore the inaccurate name I suppose :P

**Comment by mloning** [comment]
@marielledado If you're interested, you could even copy over the entire notebooks for forecasting and classification from the code breakfast, I think they are a bit more engaging than the ones we currently have.

**Comment by marielledado** [comment]
@mloning I can do that! Or rather I can copy-paste the bits that aren't already in the the existing tutorials.

**Comment by marielledado** [comment]
@mloning Pushed my final changes! I also updated my initial comment to summarize all the changes I made.

Some of the checks have already failed! Please let me know if I need to change anything. :)

**Comment by mloning** [comment]
@marielledado Yes, some tests now fail because they still use the gunpoint dataset. 

* If I remember correctly, some unit test check the accuracy on gun point, I'm not sure how to replace these. 
* The other errors should be fixed easily by replace all `load_gunpoint()` calls with either [`make_classification_problem`](https://github.com/alan-turing-institute/sktime/blob/7946e3588d0e526193be96ebd5ca250f07fb8890/sktime/utils/_testing/series_as_features.py#L33) which can be important from `sktime/utils/_testing` or loading one of the other data sets. If possible, I'd prefer `make_classification_problem()` because that generates a smaller data set which should speed up the tests somewhat.

**Comment by marielledado** [comment]
@mloning 
Appears to me as if nearly all unit tests use the gunpoint dataset:

- test_Rocket.py
- test_proximity_forest.py 
- test_pipeline.py
- test_TimeSeriesForestClassifier.py
- test_mrseql.py
- test_compose.py
- test_TSCStrategy.py
- test_FittedParamExtractor.py
- test_tasks.py
- test_orchestration.py

Could you tell me which errors would be just a matter of replacing `load_gunpoint()` with `make_classification_problem()`? Or what I have to look out for to make this replacement?

**Comment by mloning** [comment]
As far as I can tell, `test_mrseql`, `test_proximity_forest` and `test_rocket` depend on gunpoint as they check exact prediction results. The other ones should work with `make_classification_problem`.

**Comment by mloning** [comment]
Perhaps it's best to separate this work into two PRs. One for updating and replacing gun point in the tutorial notebooks and another one for trying to remove gun point altogether.

**Comment by marielledado** [comment]
@mloning good idea, in that case I will restore the gunpoint data set to complete this PR and then make a separate one with updated unit tests.

**Comment by mloning** [comment]
@marielledado Looks great, especially the shapelet notebook, good work! I've made some minor cosmetic changes and pushed them directly onto your branch. We'll merge everything once the tests pass! Thanks again! 🙌

**Comment by jasonlines** [comment]
Hi, thanks for the contribution @marielledado. 

I have no issue with replacing `GunPoint` as a sample problem in general, but I'm not keen on the proposed changes to the shapelet transform example notebook because it changes the purpose of the notebook. The reason this dataset is used in the example is because it was used in the research and it allows for interpretation of how the algorithm can be used; the original shapelet paper by Ye et al. used `GunPoint` as a case study to demonstrate that shapelets can be effective on motion problems. The dataset was then used again in the shapelet transform paper to reinforce that the transform provided additional intuitive discriminatory information that was not available from using the original algorithm (i.e. the second cluster of shapelets that is discussed in the paper and notebook). 

Changing the dataset for this notebook to `OSULeaf` unfortunately removes this discussion and interpretation, and link to the original research, as this dataset is not well-suited in the example (the accuracy in the proposed replacement workbook is only about 30%) and this negates the point of having the example notebook in my opinion. 

Can I ask what the rationale for removing `GunPoint` is? I think perhaps the origins and intention of the dataset have been misunderstood. Just to be clear, there are no actual guns used in this data, nothing is pointed at anyone, and the dataset does not glorify violence. If anything, it could be argued that this is an _anti_-violence dataset because the objective is to detect the presence of a gun (but obviously this is not a realistic real-world problem because we do not have wrist-based accelerometer data available for everyone in the world!)

We want to be inclusive and not cause offence to anyone so I am of course happy for mentions of the dataset to be removed/avoided elsewhere if this is currently the case. However, for the reasons I've mentioned above, I would prefer that the underlying data was not changed in this example and I think once proper context is given then it should be clear that this dataset does not glorify violence (but please correct me if my assumptions on this are wrong). 

As a compromise, I'd like to suggest that the following would be a better solution for this particular workbook:

1. Add extra discussion and context about what the data actually is to avoid any misunderstanding 
2. I can see that the name of the dataset could easily be a cause of confusion because it includes `gun`. We can't rename the underlying data everywhere as it is a public dataset created by an independent research group, but we could instead refer to it only as `PropPoint` in the shapelet notebook/anywhere it appears in sktime. We can also change words like "holster" to "waist" and "point" to "raise" and "lower". Would this be suitable in this example? It is completely irrelevant to the problem that the prop in question is of a gun and notionally the actor could be holding anything (for example, a lazer pointer like those used in a lecture theatre as the prop would generate very similar data) 

What are your thoughts on this? I have no issue with removing the data elsewhere in `sktime` where other data can be used interchangeably, but in this case the underlying series are important for the example.  Hopefully by explaining the problem better and removing evocative language in the example then this would clear up any misunderstandings, but I'm interested to hear your thoughts on this since you've identified an issue . And of course, I'd be happy to discuss this further if I'm missing the point as we want to make sure we don't cause offence!

**Comment by marielledado** [comment]
Hello @jasonlines thanks for your thoughts.

I suppose I really don't have to explain why a gun-related data set would make people feel uncomfortable. If you have been paying attention to the news, you would know some communities are disproportionally affected by gun violence than others. The dataset doesn't glorify guns, but it certainly normalizes it. And nowhere in the original article does it state that it's anti-violence, that's solely your interpretation and saying that would be completely insincere in my opinion. The association with guns, no matter how you spin it, will leave a bad tastes in people's mouth. 

If you genuinely "want to be inclusive and not cause offence to anyone", you would be mindful about how having a gun-related dataset would affect potential users for whom the threat of gun violence is very real. 

I spoke to @mloning before implementing any of these changes and he assured me that having a high accuracy in the tutorials is not required. My question is, why does the transform only work well with the GunPoint dataset? If it is as useful as you say it is, surely it will perform well with other datasets?

**Comment by TonyBagnall** [comment]
hi, 
"My question is, why does the transform only work well with the GunPoint dataset? If it is as useful as you say it is, surely it will perform well with other datasets?"
to answer this part, ST works on many other datasets, it is actually the best performing single representation transform, but the is no one transform that beats all others, hence why we ensemble.  Here is a link to our paper on the relative merits of different algorithms https://link.springer.com/article/10.1007/s10618-016-0483-9

**Comment by marielledado** [comment]
@jasonlines I should also mention that I attended the PyData Amsterdam code breakfast workshop. @mloning mentioned that for the workshop materials they decided to switch the Gunpoint dataset out for the arrowhead instead. It appears the team is not oblivious to the impact that this dataset might have on users. I also would have a good think about the potential ramifications of making models trained on gun-related datasets so easily accessible.

If you say that OSULeaf might not be the best dataset to illustrate ST, then I'm happy to look for a more appropriate dataset. Thanks @TonyBagnall for the link to the paper, I'll have a look.

**Comment by jasonlines** [comment]
Hi @marielledado, thank you for your replies. 

First of all I think you may have misinterpreted my intent. I can tell this is an important issue to you  which is why I reached out and I would like to discuss this rationally, professionally, and politely. Please do not use statements like "if you have been paying attention to the news" or suggesting that I am "oblivious". Statements like this are unfounded, inaccurate, and upsetting. I am, of course, very aware of current events and the very reason for me contacting you is because you raised this issue and I wanted to better understand your concerns so we can move forward appropriately. 

I would like to reiterate that I only made suggestions for how we could make this example more appropriate and I asked your opinion on that, not to justify why guns are a bad thing. This is clear to me, and obviously it is clear to you. My point was that the new dataset/workbook in this PR is not a good use case for shapelets so I was hoping to keep the underlying data the same for the use case while removing any inflammatory and inappropriate references to guns (and related terminology). I thought this would be a suitable solution and wanted to know your thoughts on this. That was my _proposed_ solution and I have no political agenda here.

To reiterate, the dataset does not contain data derived from actual firearms and there is no way that any model trained with this data could contribute to violence. Any suggestion that it could leads me to believe that you still do not understand what this dataset is because it is accelerometer data from a smart watch, **but that is fine and an important observation** - everything else aside, clearly this is still an evocative dataset to you, even after discussing it, and if it's upsetting to even one person after this explanation then clearly it is an issue that need addressing in case others feel this way.  

Going forward
=============

Hopefully with that out of the way you can now understand where I am coming from. This is not my dataset and I am not an advocate of it, but changing it without due care is detrimental to the example. It's clear that you do not like the idea of using the data, even with all references to firearms removed, so I'm happy for another dataset to be used in place of it in this example (and I have no problem with GunPoint being removed everywhere else like I've already mentioned). But clearly OSULeaf isn't a good fit for this example with shapelets. You are correct that accuracy isn't everything, and we don't need one that is 100% for the notebook example, but I would like it to include an example dataset that demonstrates how the approach can produce interpretable features like the previous iteration did, rather than simply just reporting how to use it.

What I'd propose for now would be to just delete the notebook as part of your PR and we can reinstate it later. I'm more than happy for you to propose new data to be used in its eventual replacement and for me to prepare it, or you can go ahead and author it yourself if you like, but I'd like to be included in the process to make sure that we get the example right (@mloning has not been involved in any time series classification or shapelet research so I wouldn't expect him to be able to comment on this!). 

Does this sound like a good and suitable way forward to you?

**Comment by jasonlines** [comment]
(or alternatively we could update it with the OSULeaf dataset as-is for now, and then find a nicer example later in another issue?)

**Comment by marielledado** [comment]
@jasonlines I can see how my comment upset you. If my tone offended you, I apologize (although you telling me that I "do not understand" the data set is not what I would consider discussing this issue "rationally, professionally, and politely" either)

For my part, I was frustrated at what I perceived as having to justify to you why I felt strongly enough about replacing the GunPoint dataset to make a pull request. I got this impression from because your proposed suggestion was to cover up the gun-related terms. This would not be enough, as it would just be a "re-package" of the thing that's actually inappropriate, which is the dataset's association with guns. What I had hoped for in your initial response to the PR was to simply point out that OSULeaf is not appropriate for ST and maybe a conversation on exploring other datasets (for example, http://www.timeseriesclassification.com/ has whole section on motion data, I'm sure we can find something there). 

I don't know how else to interpret a question "Can I ask what the rationale for removing GunPoint is?", other than soliciting a justification. The thing is, you said things that lead me to believe that you had an idea why I was asking for this change: you said no real firearms were involved, that it could be interpreted as anti-violence etc, as if to reason with me about how it's not as bad as I think it is. Combine all that with your suggestion to simply change some terms, it was impossible for me to not to read whole thing as a defense of the Gunpoint dataset.

As for moving forward, you are welcome to merge or discard these changes as you see fit. As ST is not my field of expertise I don't suppose I can be of much help.

**Comment by fkiraly** [comment]
@jasonlines , to give some supplementary information about the workshop: in preparation of the workshop notebooks, I indeed suggested strongly to remove the gunpoint dataset from it, given current events in the US, and also in general (having thought about it).

The dataset is about detecting whether someone points a gun (or just pretends).
The act itself, and shown illustratively in what is more or less one of the first example notebooks, may cause discomfort in people, especially those with experience of weapon violence.
While I personally have been less troubled by that, I know that others may feel more strongly.

For me, it was more the application which left quite an Orwellian impression when I saw it.
Ramifications in surveillance or military scenarios are easy to come up with.

From a scientific perspective, I agree with @jasonlines that purging the "toy dataset that everyone in a certain community is using" in its entirety from sktime might not be a good way forward, e.g., for backwards compatibility, automated testing, and reproducibility of results. I also don´t necessarily think that it is always unethical to collate or work on a dataset like this, e.g., if you think about automated protection mechanisms at a bank or airport - which is still close to ethical boundaries but can still be seen on the right side of them (right next to a couple of very slippery slopes).

Anyway, what made me personally suggest to not have it in the workshop tutorials very much echoes the sentiments of @marielledado, who, independently of my recommendation, suggested to refactor the sktime native tutorials when she was working through them. 
Based on the above, and being considerate of our users, I personally think it is somewhat in bad taste to have it showcased that prominently, illustration of gun pointing and all...

**Comment by fkiraly** [comment]
@jasonlines , on a purely methodological side: why is OSUleaf not appropriate for shapelet transform? As far as I understand it, shapelet transform should address any problem where you would expect that an index localized change in the sequence is discriminative. 

I´m not a plant biologist, but I would find this to be plausible to happen in leaves, compare for example obovate vs obcordate vs truncate here:
https://www.ck12.org/c/biology/leaf-types/lesson/Leaf-Types-Advanced-BIO-ADV/

Anyway, a typical application of shapelets in a health/medical context would be ECG analysis, right? Because the morphology that discriminates physiological from pathological is localized in index (time). Do we have a dataset like that?

**Comment by ABostrom** [comment]
hi @fkiraly, thought I'd chime in on why OSULeaf isn't as good a dataset at first glance. 

image

I've put a picture of some of the leaves from OSULeaf. The types of algorithms you would expect to be most performant on leaf type data is dictionary based algorithms, I believe BOSS has the highest accuracy. Whilst your assumption that the shapelets capture the "shape" information inherent in the outline of the leaf. Leaves tend to be symmetrical, and so often the number of repeating motifs is more discriminatory than just the shape itself.

**Comment by fkiraly** [comment]
@abostrom, stupid question about OSUleaf - this all looks like the same kind of leaf to me? ... palmate?

**Comment by ABostrom** [comment]
@fkiraly Yes you're right. I believe that leaf i linked is actually just the species Acer Circinatum. 
The point I was trying to make is that repeated motifs is the feature that is best captrued, and so OSULeaf is probably best used on a dictionary example. In the dataset i think four of the six classes are from the Acer family though, so they are deliberately similar. 

This is a better link to the leaves present in the dataset.
[OSULeaf website](http://web.engr.oregonstate.edu/~tgd/leaves/dataset.htm)

This shows how the outline is generated, but does demonstrate how the series has similar motifs spread throughout. 
[Leaf Outline Generation](http://web.engr.oregonstate.edu/~tgd/leaves/process.htm)

**Comment by jasonlines** [comment]
Hi @marielledado, I thought it was best to step away for a few days as I didn't want to say anything else in case I inadvertently made the situation worse! I won't comment on the details of this PR any further but I wanted to apologise if my tone or choice of language upset you too. It was never my intention to start an argument, or be patronising or unwelcoming, but it's clear that was the result no matter the intent and it's my responsibility to acknowledge that. So for that I am sorry, and I'll make an effort to be more considerate in my future choice of language. And sincerely, thank you again for your contributions in this PR. I hope my bumbling of this notebook discussion doesn't put you off contributing to sktime and/or other open-source communities in the future

**Comment by marielledado** [comment]
@jasonlines Thank you, I accept your apology. Moving forward I will also be more considerate about my words, regardless of my feelings about a particular issue. I will definitely continue contributing to OS projects!

**Comment by mloning** [comment]
I'll merge this PR now. Thanks again @marielledado for your work on this and sorry for the lack of communication between me and the other core devs.  

I've opened a separate issue #320  to find a more suitable data set to illustrate the shapelet transform.
